### ISSUE TITLE
How to create and train model from dataset

### ISSUE THREAD

**Issue by Di3GO95** [stanfordnlp/CoreNLP#921]
Hello, I'm trying to create a sentiment analysis app in java, to retrieve tweets and analize them. I've already made it so that it works with the default model. 

Now I want to go one step further and create my own model. I have searched tutorials but I still don't understand how to do it, in most cases these tutorials just use the default models.

In my case, I already have 3 XML files, "training.xml", "development.xml" and "test.xml", that have the following structure, with 4 possible sentiments (P, N, NEU, NONE):

```
<tweets>
   <tweet>
      (tweet info, username, id...)
      <content>tweet content</content>
      <sentiment>
         <polarity>
            <value>N</value>
         </polarity>
      </sentiment>
   </tweet>
   (...)
</tweets>
```

But at this point, I don't know what are the steps convert these XML to a training model. 

Thank you for your time.

**Comment by AngledLuffa** [comment]
Are you asking for help parsing the XML or asking for the output format?

**Comment by Di3GO95** [comment]
> Are you asking for help parsing the XML or asking for the output format?

I guess that the question would be more like "what are the steps". I don't know how to analize and modify these datasets so that the model is created from them (I guess that I have to modify them first, in a way that Stanford NLP "understands" these XML, but what modifications are required?), but I'm not sure what are the exact steps to go from these datasets to the model to use in my program...  

So what I'm asking, would be something like:

- You have those training, development and test XML files,  so... to create a model to use in NLP algorithm you have to do:
1) ...
2) ...
3) ...

- Or follow this tutorial: (link). But after searching, I only find tutorials where they use the models of the official website.

**Comment by AngledLuffa** [comment]
Take a look at BuildBinarizedDataset:

https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/sentiment/BuildBinarizedDataset.java


What you need to do is convert the XML to the input for this tool, then run
the tool to produce trees such as the sentiment trainer can process.

John

**Comment by Di3GO95** [comment]
Thank you for your answer. I have one question, it says that after the line of the text (in my case, the tweet), there will be the sub-phrases, and then a blank line to separate this block from the next one. Do I need to do this for every tweet that I have, or will it work if it's just the whole tweet with its sentiment?

> Take a look at BuildBinarizedDataset: https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/sentiment/BuildBinarizedDataset.java What you need to do is convert the XML to the input for this tool, then run the tool to produce trees such as the sentiment trainer can process. John

**Comment by AngledLuffa** [comment]
It's better if you have subphrases (hence the RNTN aspect of the model) but
it will work without them

**Comment by Di3GO95** [comment]
One last question, I have created my model and it has generated multiple files (model.ser, model-0000-41,26.ser, model-0001-40,07.ser, ...). I guess that all of them are my model (though I expected just 1 file). 

In my code, I have:

```
props = new Properties();
props.setProperty("annotators", "tokenize, ssplit, parse, sentiment");
pipeline = new StanfordCoreNLP(props);
```
So I guess that I have to pass the path to my models.ser file, right? What is the right property to set the path of my model?

Thank you again for your patience.

**Comment by AngledLuffa** [comment]
It's just archiving a bunch of intermediate models.  The path you want is
whichever one has the best dev score

**Comment by Di3GO95** [comment]
Ok, so I guess that the score is the number next to the index of the file, right? For example, model-0000-41,26.ser would have better score (41.26) than model-0001-40,07.ser (40,07), so between these two, I would choose the first one, right? 

If this is right and that number is the score, why doesn't the first file (model.ser) have any score?

And the code to load that model is this. if I am not wrong?

```
props = new Properties();
props.setProperty("annotators", "tokenize, ssplit, parse, sentiment");
props.setProperty("parse.model", path_to_my_custom_model);
pipeline = new StanfordCoreNLP(props);
```

Thank you.

**Comment by AngledLuffa** [comment]
I believe that should work.  Also, I can't remember exactly, but it's
possible that model.ser is the highest scoring model.  You can probably
figure it out with

ls -l

**Comment by Di3GO95** [comment]
I have tried to add the property like this:

```
props = new Properties();
props.setProperty("annotators", "tokenize, ssplit, parse, sentiment");
String file = "stanford/model.ser.gz";
String model = getClass().getClassLoader().getResource(file).getFile(); 
// It finds the file correctly
props.setProperty("sentiment.model", model);
pipeline = new StanfordCoreNLP(props);
```

But when I try to run it (in eclipse, not in command line), i get the error 
```
Caused by: java.lang.ClassNotFoundException: com.SentimentModel
```
I've not had any problem up until now, using the default model through Maven:

```
<dependency>
	<groupId>edu.stanford.nlp</groupId>
	<artifactId>stanford-corenlp</artifactId>
	<version>3.9.2</version>
</dependency>
<dependency>
	<groupId>edu.stanford.nlp</groupId>
	<artifactId>stanford-corenlp</artifactId>
	<version>3.9.2</version>
	<classifier>models</classifier>
</dependency>
```

What should I do to get rid of this problem? I thought that even if I used a custom model, there wouldn't be any problem with the dependencies.

Thank you.

**Comment by AngledLuffa** [comment]
Are you saying it works on the command line and not in eclipse?

**Comment by Di3GO95** [comment]
No, it's that I have only tried to run the code in eclipse.

Looking at:

```
Caused by: java.lang.ClassNotFoundException: com.SentimentModel
```

Maybe the problem is that I just copied the folder edu\stanford\nlp\sentiment of CoreNLP-master to a new project in eclipse, to run it there instead of using command line, and generate the train and dev trees, and the model. And the package where these classes were placed was named "com". So I think that this is the problem, probably. So I'm going to try to to generate the trees and the model through command line in the CoreNLP-master that I have downloaded, instead of my own project with my packages.

**Comment by Di3GO95** [comment]
It was that, now there is no error. But now there is something strange... I generated two models, with different amount of training and dev data. When I use the model with less data it works fine, but when I use the one generated with more data, the program runs fine but it's always the same output:

```
very positive: 0.0%
positive: 0.0%
neutral: 0.0%
negative: 0.0%
very negative: 0.0%
```

As I said in previous comments, the property has been modified with the next line, so It's probably that I messed up generating the trees or the model, and not a code problem, right?

```
props.setProperty("sentiment.model", model);
```

And one little thing that's weird (at least to me), is that the model generated with less data, has a size of 1477KB, while the one generated with more data, just 126kb... It's weird, I thought that the size would be much bigger than the first one, it even took 5 more hours to be generated than the model with less data...
